File: src/solutions/2025/07/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" false;;
Hashtbl.add options "has_io" false;;

let solve (input : string) (_ : bool) (p2 : bool) : string =
  let total = ref 0 in
  let len = String.index input '\n' in
  let buffer = Array.make 141 0 in
  buffer.(String.index input 'S') <- 1;
  for i = 0 to String.length input - 1 do
    if input.[i] == '^'
    then (
      let x = i mod (len + 1) in
      if buffer.(x) > 0 then total := !total + 1;
      buffer.(x - 1) <- buffer.(x - 1) + buffer.(x);
      buffer.(x + 1) <- buffer.(x + 1) + buffer.(x);
      buffer.(x) <- 0)
  done;
  if p2 then string_of_int (Array.fold_left ( + ) 0 buffer) else string_of_int !total
;;

let part_1 (input : string) (_is_test : bool) : string = solve input true false
let part_2 (input : string) (_is_test : bool) : string = solve input true true
let () = Aoc.run Sys.argv part_1 part_2 options

File: src/solutions/2025/07/lua/main.lua
local run = require('src.langs.lua.run')

local options = {
   has_alternate = false,
   has_io = false,
}

---@param input string
---@param is_test boolean
---@param p2 boolean
---@return string
local function solve(input, is_test, p2)
   local total = 0
   local len = input:find('\n')
   local buffer = {}
   for i = 1, len - 1 do
      buffer[i] = 0
   end
   buffer[input:find('S')] = 1

   for i = 1, #input do
      local c = input:sub(i, i)
      if c == '^' then
         local x = i % len
         if buffer[x] > 0 then
            total = total + 1
         end
         buffer[x - 1] = buffer[x-1] + buffer[x]
         buffer[x + 1] = buffer[x+1] + buffer[x]
         buffer[x] = 0
      end
   end

   if p2 then
      total = 0
      for _, v in pairs(buffer) do
         total = total + v
      end
   end
   
   return tostring(total)
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   return solve(input, is_test, false)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   return solve(input, is_test, true)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()